### DISCUSSION TITLE
Links extraction

### DISCUSSION THREAD

**Discussion by meghanaviyyapu** [pymupdf/PyMuPDF#3294] [Looking for help] [answered]
I have tried extracting link text from the below PDF. Along with the link text was able to view additional text.
Have used page.get_textbox(rect) to extract link text. Could you let me know how only link text can be obtained
[GUL_link.pdf](https://github.com/pymupdf/PyMuPDF/files/14720491/GUL_link.pdf)

**Comment by JorjMcKie** [comment]
Your PDF creator unfortunately uses a too small inter-line distance. So line bboxes overlap each other:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/b1c0d0ca-bc04-4b5a-bc3a-d279894664a8)
Method `page.get_textbox()` is optimized for including any character that overlaps the desired rectangle. It therefore is not a good candidate here.
Using `page.get_text(clip=link["from"])` is better but still includes "y" and "g" - the reason can be seen in the image.
```python
for link in page.get_links():
    print(page.get_text(clip=link["from"]).replace("\n",""))

    
www.MetLife.com
ygHIPAAprivacyAmericasUS@metlife.com
```
You could experiment a little and reduce the height of the link rect a bit from its top:
```python
for link in page.get_links():
    print(page.get_text(clip=link["from"]+(0,2,0,0)))

    
www.MetLife.com

HIPAAprivacyAmericasUS@metlife.com
```
This is much better.

**Reply by meghanaviyyapu** [reply]
Thanks for explaining. How to find out which value to be added to clip=link["from"] because in other pdfs the same value may or may not work.

**Reply by JorjMcKie** [reply]
Difficult / hard to predict! Because there so many sloppy PDF creators!
You could change your strategy somewhat:
1. make a list of words from the full page
2. when iteration over the `link["from"]` rectangles, select words that **_"mostly_** are contained in the link rectangle. Something like this:

```python
# prepare list of word rectangles and word text
words=[(fitz.Rect(w[:4]), w[4]) for w in page.get_text("words")]

# iterate over the links and select words whose rectangles are inside by at last 80%:
for link in page.get_links():
    frect=link["from"]
    link_area = 0.8 * abs(frect)  # area of link rect
    for w, text in words:  # check word intersection size
        if abs(w & frect) >= link_area:
            print(text)

            
www.MetLife.com.
HIPAAprivacyAmericasUS@metlife.com,
```

**Reply by JorjMcKie** [reply]
To get rid of the trailing punctuations use `punct = tuple(string.punctuation)` and replace the text by `text = text[:-1] if text.endswith(punct) else text`.

**Reply by meghanaviyyapu** [reply]
Thanks a lot for the detailed examples

**Reply by meghanaviyyapu** [reply]
I have tried extracting links from the below PDF
[BTA_link.pdf](https://github.com/pymupdf/PyMuPDF/files/14753151/BTA_link.pdf)
 The link got identified as "**CERTIFICATE FACE PAGE**" 
 When I tried extracting spans, it got extracted as **CERTIFICATE PAG** and **E..........**. 
 
 Could you let me know if it is possible to get the span output identical to link output like CERTIFICATE PAGE instead of CERTIFICATE PAG?

**Reply by JorjMcKie** [reply]
The span looks like this:
```python
pprint(spans[3])
{'ascender': 0.9052734375,
 'bbox': (72.0, 84.2972412109375, 560.7783203125, 95.4691162109375),
 'color': 0,
 'descender': -0.2119140625,
 'flags': 4,
 'font': 'ArialMT',
 'origin': (72.0, 93.3499755859375),
 'size': 10.0,
 'text': 'CERTIFICATE FACE PAGE............................................................................................................................... '
         '2 '}
```
And thus covers the full line.
But there are **_3 different links_** covering this same line:
![image](https://github.com/pymupdf/PyMuPDF/assets/8290722/1ca587ae-e58d-455a-a18c-80ace563be8f)

IAW for the text of the span **_you will not find_** a link that covers it completely.
So you somehow have to deal with this by adjusting your code appropriately ...

**Comment by meghanaviyyapu** [comment]
Thanks, is there a way to find out if the same span contains multiple links?

**Reply by JorjMcKie** [reply, marked as answer]
Look at the red rectangles! These are the "from" rectangles of the links.
